File: mobile-sdks/android/yext-analytics/src/main/java/com/yext/analytics/AnalyticsEventReporter.kt
package com.yext.analytics

import android.content.Context
import android.os.Build

/**
 * Main analytics event reporter implementing the AnalyticsEventService interface
 */
class AnalyticsEventReporter private constructor(
    private val config: AnalyticsConfig,
    private val context: Context,
    private val defaultPayload: EventPayload? = null
) : AnalyticsEventService {
    
    private val sessionManager = SessionManager(context)
    
    companion object {
        private const val SDK_VERSION = "1.0.0"
        
        /**
         * Create a new AnalyticsEventReporter instance
         */
        fun create(config: AnalyticsConfig, context: Context): AnalyticsEventReporter {
            return AnalyticsEventReporter(config, context)
        }
    }
    
    /**
     * Create a new analytics service with additional default payload values
     */
    override fun with(payload: EventPayload): AnalyticsEventService {
        val mergedPayload = merge(defaultPayload, payload)
        return AnalyticsEventReporter(config, context, mergedPayload)
    }
    
    /**
     * Report an analytics event
     */
    override suspend fun report(newPayload: EventPayload?): String {
        val finalPayload = merge(defaultPayload, newPayload ?: EventPayload())
        
        // Handle session tracking
        if (!config.sessionTrackingEnabled) {
            finalPayload.sessionId = null
        } else if (finalPayload.sessionId == null) {
            finalPayload.sessionId = sessionManager.getOrCreateSessionId()
        }
        
        // Add client SDK version
        if (finalPayload.clientSdk == null) {
            finalPayload.clientSdk = mutableMapOf()
        }
        finalPayload.clientSdk!!["ANALYTICS"] = SDK_VERSION
        
        // Set authorization
        finalPayload.authorization = when (config.authorizationType) {
            AnalyticsConfig.AuthorizationType.API_KEY -> "KEY ${config.authorization}"
            AnalyticsConfig.AuthorizationType.BEARER -> "Bearer ${config.authorization}"
        }
        
        // Get device information if not provided
        if (finalPayload.browserAgent == null) {
            finalPayload.browserAgent = getDeviceInfo()
        }
        
        // Setup request URL
        val url = NetworkManager.setupRequestUrl(config.environment, config.region)
        
        // Send the event
        return NetworkManager.sendEvent(url, finalPayload, config)
    }
    
    /**
     * Merge two event payloads, with new payload taking precedence
     */
    private fun merge(base: EventPayload?, new: EventPayload): EventPayload {
        if (base == null) return new.copy()
        
        return base.copy().apply {
            new.action?.let { action = it }
            new.authorization?.let { authorization = it }
            new.bot?.let { bot = it }
            new.browserAgent?.let { browserAgent = it }
            new.chat?.let { chat = it }
            new.clientSdk?.let { clientSdk = mergeStringMaps(clientSdk, it) }
            new.count?.let { count = it }
            new.customTags?.let { customTags = mergeStringMaps(customTags, it) }
            new.customValues?.let { customValues = mergeDoubleMaps(customValues, it) }
            new.destinationUrl?.let { destinationUrl = it }
            new.entity?.let { entity = it }
            new.internalUser?.let { internalUser = it }
            new.ip?.let { ip = it }
            new.label?.let { label = it }
            new.locale?.let { locale = it }
            new.location?.let { location = it }
            new.pages?.let { pages = it }
            new.pageUrl?.let { pageUrl = it }
            new.referrerUrl?.let { referrerUrl = it }
            new.search?.let { search = it }
            new.searchTerm?.let { searchTerm = it }
            new.sessionId?.let { sessionId = it }
            new.timestamp?.let { timestamp = it }
            new.value?.let { value = it }
            new.visitor?.let { visitor = mergeStringMaps(visitor, it) }
        }
    }
    
    private fun mergeStringMaps(base: Map<String, String>?, new: Map<String, String>): Map<String, String> {
        return (base ?: emptyMap()) + new
    }
    
    private fun mergeStringMaps(base: MutableMap<String, String>?, new: Map<String, String>): MutableMap<String, String> {
        val result = base?.toMutableMap() ?: mutableMapOf()
        result.putAll(new)
        return result
    }
    
    private fun mergeDoubleMaps(base: Map<String, Double>?, new: Map<String, Double>): Map<String, Double> {
        return (base ?: emptyMap()) + new
    }
    
    private fun getDeviceInfo(): BrowserAgent {
        return BrowserAgent(
            browser = "Mobile App",
            browserVersion = SDK_VERSION,
            device = "${Build.MANUFACTURER} ${Build.MODEL}",
            deviceClass = getDeviceClass(),
            os = "Android",
            osVersion = Build.VERSION.RELEASE,
            userAgent = "YextAnalytics Android SDK $SDK_VERSION"
        )
    }
    
    private fun getDeviceClass(): String {
        return when {
            context.resources.configuration.smallestScreenWidthDp >= 600 -> "tablet"
            else -> "phone"
        }
    }
}

File: mobile-sdks/ios/YextAnalytics/Sources/YextAnalytics/AnalyticsEventReporter.swift
import Foundation

/// Main analytics event reporter implementing the AnalyticsEventService protocol
public class AnalyticsEventReporter: AnalyticsEventService {
    private let config: AnalyticsConfig
    private let defaultPayload: EventPayload?
    
    /// Initialize the analytics reporter with configuration
    public init(config: AnalyticsConfig, defaultPayload: EventPayload? = nil) throws {
        // Validate configuration
        guard !config.authorization.isEmpty else {
            throw AnalyticsError.invalidConfiguration("Authorization must be provided")
        }
        
        self.config = config
        self.defaultPayload = defaultPayload
    }
    
    /// Create a new analytics service with additional default payload values
    public func with(_ payload: EventPayload) -> AnalyticsEventService {
        let mergedPayload = merge(base: defaultPayload, new: payload)
        return try! AnalyticsEventReporter(config: config, defaultPayload: mergedPayload)
    }
    
    /// Report an analytics event
    public func report(_ newPayload: EventPayload? = nil) async throws -> String {
        var finalPayload = merge(base: defaultPayload, new: newPayload)
        
        // Handle session tracking
        if !config.sessionTrackingEnabled {
            finalPayload.sessionId = nil
        } else if finalPayload.sessionId == nil {
            finalPayload.sessionId = SessionManager.getOrCreateSessionId()
        }
        
        // Add client SDK version
        let version = getSDKVersion()
        if finalPayload.clientSdk == nil {
            finalPayload.clientSdk = [:]
        }
        finalPayload.clientSdk?["ANALYTICS"] = version
        
        // Set authorization
        finalPayload.authorization = config.authorizationType == .apiKey
            ? "KEY \(config.authorization)"
            : "Bearer \(config.authorization)"
        
        // Get device information if not provided
        if finalPayload.browserAgent == nil {
            finalPayload.browserAgent = getDeviceInfo()
        }
        
        // Setup request URL
        let url = try NetworkManager.setupRequestUrl(environment: config.environment, region: config.region)
        
        // Send the event
        return try await NetworkManager.sendEvent(to: url, payload: finalPayload, config: config)
    }
    
    /// Merge two event payloads, with new payload taking precedence
    private func merge(base: EventPayload?, new: EventPayload?) -> EventPayload {
        guard let base = base else { return new ?? EventPayload() }
        guard let new = new else { return base }
        
        var merged = base
        
        // Merge all properties, with new taking precedence over base
        if new.action != nil { merged.action = new.action }
        if new.authorization != nil { merged.authorization = new.authorization }
        if new.bot != nil { merged.bot = new.bot }
        if new.browserAgent != nil { merged.browserAgent = new.browserAgent }
        if new.chat != nil { merged.chat = new.chat }
        if new.clientSdk != nil { merged.clientSdk = mergeStringDictionaries(base.clientSdk, new.clientSdk) }
        if new.count != nil { merged.count = new.count }
        if new.customTags != nil { merged.customTags = mergeStringDictionaries(base.customTags, new.customTags) }
        if new.customValues != nil { merged.customValues = mergeNumberDictionaries(base.customValues, new.customValues) }
        if new.destinationUrl != nil { merged.destinationUrl = new.destinationUrl }
        if new.entity != nil { merged.entity = new.entity }
        if new.internalUser != nil { merged.internalUser = new.internalUser }
        if new.ip != nil { merged.ip = new.ip }
        if new.label != nil { merged.label = new.label }
        if new.locale != nil { merged.locale = new.locale }
        if new.location != nil { merged.location = new.location }
        if new.pages != nil { merged.pages = new.pages }
        if new.pageUrl != nil { merged.pageUrl = new.pageUrl }
        if new.referrerUrl != nil { merged.referrerUrl = new.referrerUrl }
        if new.search != nil { merged.search = new.search }
        if new.searchTerm != nil { merged.searchTerm = new.searchTerm }
        if new.sessionId != nil { merged.sessionId = new.sessionId }
        if new.timestamp != nil { merged.timestamp = new.timestamp }
        if new.value != nil { merged.value = new.value }
        if new.visitor != nil { merged.visitor = mergeStringDictionaries(base.visitor, new.visitor) }
        
        return merged
    }
    
    private func mergeStringDictionaries(_ base: [String: String]?, _ new: [String: String]?) -> [String: String]? {
        guard let base = base else { return new }
        guard let new = new else { return base }
        return base.merging(new) { _, new in new }
    }
    
    private func mergeNumberDictionaries(_ base: [String: Double]?, _ new: [String: Double]?) -> [String: Double]? {
        guard let base = base else { return new }
        guard let new = new else { return base }
        return base.merging(new) { _, new in new }
    }
    
    private func getSDKVersion() -> String {
        return "1.0.0" // This should be read from the package info in a real implementation
    }
    
    private func getDeviceInfo() -> BrowserAgent {
        let device = UIDevice.current
        let osVersion = device.systemVersion
        let deviceModel = getDeviceModel()
        
        return BrowserAgent(
            browser: "Mobile App",
            browserVersion: getSDKVersion(),
            device: deviceModel,
            deviceClass: getDeviceClass(),
            os: "iOS",
            osVersion: osVersion,
            userAgent: "YextAnalytics iOS SDK \(getSDKVersion())"
        )
    }
    
    private func getDeviceModel() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value))!)
        }
        return identifier
    }
    
    private func getDeviceClass() -> String {
        switch UIDevice.current.userInterfaceIdiom {
        case .phone:
            return "phone"
        case .pad:
            return "tablet"
        case .tv:
            return "tv"
        case .carPlay:
            return "car"
        case .mac:
            return "desktop"
        case .vision:
            return "headset"
        default:
            return "unknown"
        }
    }
}

#if canImport(UIKit)
import UIKit
#endif